posts/inside-rust/2021-08-27-Splitting-const-generics.md

```diff
@@ -0,0 +1,124 @@
+---
+layout: post
+title: "Splitting the const generics features"
+author: lcnr
+description: "Splitting the const generics features"
+team: The Const Generics Project Group <https://rust-lang.github.io/project-const-generics/>
+---
+
+After the stabilization of the const generics MVP in version 1.51, the const generics project group has continued to
+work on const generics. Large parts of this work were gated behind the feature gates `const_generics` and `const_evaluatable_checked`. As time went on, the
+`const_generics` feature became fairly useless on its own while the name of
+`const_evaluatable_checked` didn't really capture what this feature was intended to do.
+
+To improve this, we have removed the features `const_generics`, `lazy_normalization_consts`, and `const_evaluatable_checked`. They have been replaced by `feature(const_param_types)` and `feature(generic_const_exprs)`.
+
+As there is a lot going on with const generics, here's a quick overview of the new - and preexisting - features and how much still needs to be done for them to get stabilized:
+
+### `feature(const_param_types)`
+
+On stable, only integers, `char` and `bool` are allowed as the types of const parameters. This feature allows additional types, such as `&'static str` and user defined types.
+```rust
+#![feature(const_param_types)]
+
+#[derive(PartialEq, Eq)]
+enum ImageFormat {
+    Rgb8,
+    Rgba8,
+    // ...c
+}
+
+struct Image<const FORMAT: ImageFormat> {
+    // ...
+}
+
+impl Image<{ ImageFormat::Rgba }> {
+    fn alpha(&self, pixel: PixelLocation) -> u8 {
+        // ...
+    }
+}
+```
+Note that even with this feature generic const parameter types, such as `struct Foo<T, const N: T> { ... }` are forbidden.
+While allowing such things is desired, it adds additional complications exceeding our current capacity.
```

Comment:
> I guess implementation-wise, routing Bar from Foo<Bar, {expr}> into the argument to type-check it as expr: Bar, is going to be a lot of the work necessary?

yeah, we could change https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.WithOptConstParam.html to store the `typecheck_type` instead of a param def id but that seems fragile to me